Add precision and time unit to History Stats integration

cgomesu opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Add precision (`precision:`) and time unit (`time_unit:`) configuration variables to control how the state value of a `history_stats` sensor is displayed:

- `precision` indicates the number of decimals to be used in the `round()` operation prior to displaying the state value.  It affects both `type: time` and `type: ratio` sensors. The default value is `2`.
- `time_unit` indicates the unit of time (`d`, `h`, `min`, `s`) to be used for the state value of a `type: time` sensor.  Valid units are taken from the HASS [const.py](https://github.com/home-assistant/core/blob/dev/homeassistant/const.py).

The addition of the two configuration variables is motivated by [issue#52251](https://github.com/home-assistant/core/issues/52251).  More specifically:

1. Use of `hours` as time unit for the `state` value is arbitrary;
2. In the user's example, the `value` attribute shows `9m` but the `state` (`0.17` hours) suggests it should be `10m` instead (`0.17 * 60` = `10` minutes and `12` seconds). This is related to `divmod()` usage in the `pretty_duration()` static method. Specifically, if `hours < 0`, `pretty_duration()` returns only the *integer* `minutes`, which is lower than or equal to `hours*60`. However, the `state` of the entity is shown as a *rounded* (`2` decimals) *floating number* in `hours` units. Therefore, if `hours` was `.166` prior to rounding, for example, `pretty_duration()` would return `9m` because `divmod(int(0.166*3600), 60) = (9, 57)` and `minutes` is defined as `a // b` (floor division) of `divmod(a, b)` but the `state` of the entity would be `0.17` due to `round(0.166, 2)`. 

This PR fixes both issues via the inclusion of user-specified decimals (`precision`) and unit of time (`time_unit`) while maintaining backward compatibility--that is, existing configurations that make use of the History Stats integration will not be affected by these changes, owing to the default values being the same as in the current implementation of the History Stats integration.

Other noteworthy observations:

- Tests for the History Stats sensor ([`tests/components/history_stats/test_sensor.py`](https://github.com/home-assistant/core/blob/dev/homeassistant/components/history_stats/sensor.py)) were updated to include reference to the two new configuration variables. In addition, a new test sensor was created for two test methods that make use of a non-default `time_unit` (`"min"`).  New tests that specifically test `precision` and `time_unit` were not included.

- Merging this PR requires editing the docs for the History Stats integration to reference the new configuration variables.  *Edited*: Added reference to the doc PR.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #52251
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/18380

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

cgomesu committed:
Add precision configuration variable

Precision refers to the number of decimals in the state value for
a history_stats sensor of type time or ratio.

Included to fix misleading state values, owning to rounding operations
and div_mod() usage in history_stats sensors of type time
(see issue#52251).

The precision config variable takes any integer
equal or higher than 0 (zero). By default, it uses 2 decimals,
as in the prior code for type time. For type ratio, it now
outputs an additional decimal by default (e.g., 95.58 instad of
95.6). Type count is unaffected.

cgomesu committed:
Add refence to precision config variable to tests

Reference to a new precision configuration variable was added to
existing tests for the history_stats sensor.

New tests were not created to specifically test precision.

cgomesu committed:
Add time_unit config variable for type time

time_unit indicates the unit of time to be used when displaying the
amount of time in the state of a history_stats sensor of type time.
It can be days, hours (default), minutes, or seconds. A valid unit is
imported from the HASS const.py.

Units are converted via a new static method called convert_hours(),
which is called prior to returning the state of the sensor via state().

The change fix issue#52251 by giving users an option to change the
time unit of the state value.

cgomesu committed:
Add time_unit reference to tests

Updated tests for the sensor with the new time_unit configuration
variable.

Added a 5th sensor to test_measure_multiple() and
async_test_measure() to assert state value for a nondefault
time unit (min).

Ongy commented on homeassistant/components/history_stats/sensor.py:
```diff
@@ -115,7 +149,11 @@ def __init__(
+        self._precision = precision
+        self._time_unit = time_unit
+        self._unit_of_measurement = (
+            UNITS[sensor_type] if not sensor_type == CONF_TYPE_TIME else time_unit
```
I feel like a quick comment wouldn't hurt here to explain what's going on.

cgomesu replied:
Sounds reasonable.  I'll add a comment to explain the handling of units.

Ongy commented on tests/components/history_stats/test_sensor.py:
```diff
@@ -427,6 +463,7 @@ async def async_test_measure(hass):
     assert hass.states.get("sensor.sensor2").state == STATE_UNKNOWN
     assert hass.states.get("sensor.sensor3").state == "2"
     assert hass.states.get("sensor.sensor4").state == "50.0"
+    assert hass.states.get("sensor.sensor5").state == "30.0"
```
Is the precision property ever really tested?

I feel like a new test here should make use of it at least once, to make sure it does something.
But the old ones (with precision 2) didn't have to get their result changed, and new test has precision one.

cgomesu replied:
I did not write new tests for the new configuration variables, as pointed out in the PR:

> - Tests for the History Stats sensor (tests/components/history_stats/test_sensor.py) were updated to include reference to the two new configuration variables. In addition, a new test sensor was created for two test methods that make use of a non-default time_unit ("min"). New tests that specifically test precision and time_unit were not included.

I added an extra sensor (`sensor.sensor5`) to two of the existing methods to evaluate a non-default `time_unit`, namely `min` (minutes).  However, as you correctly noted, `precision` does not play any role there because the evaluated time window (half an hour) can be expressed as `30` minutes (the time unit tested) or `1800` seconds.  On a second thought, I could change the `time_unit` for `d` (days) instead of `min`, which would give us a value of `0.0125` and then also test precision (for `3`, the output should `0.013`).  I'll make this change, per your recommendation.

Ongy reviewed (commented):
Main code looks good to me though one line feels a bit harder to read than good code should.

Also feel like the new tests don't test part of the new features

cgomesu committed:
Improve readability of unit_of_measurement

Add comment about the unit_of_measurement logic

Improve code readability of unit_of_measuremente by
changing 'if not ... ==' for 'if ... !='

cgomesu committed:
Improve precision and time_unit evaluation

Change precision (1 to 3) and time_unit (minutes to days) for the
sensor.sensor5 to allow proper evaluation of both precision and
time_unit.

dgomes commented on homeassistant/components/history_stats/sensor.py:
```diff
@@ -67,6 +74,12 @@ def exactly_two_period_keys(conf):
             vol.Optional(CONF_DURATION): cv.time_period,
             vol.Optional(CONF_TYPE, default=CONF_TYPE_TIME): vol.In(CONF_TYPE_KEYS),
```
I believe you broke this line when CONF_TYPE_TIME became a list

This will impact sensory_type further below

cgomesu replied:
Thanks for chiming in, @dgomes . I'll address each of your concerns separately.

Please notice that `CONF_TYPE_TIME` **is not** a list. It is actually a string (namely, `time`), as defined in L41 of the same file. I think you got confused with its usage in the `UNITS` dict, in which it shows up as a key that has a list as value.

dgomes commented on homeassistant/components/history_stats/sensor.py:
```diff
@@ -344,9 +391,7 @@ def pretty_ratio(value, period):
-
-        ratio = 100 * 3600 * value / (period[1] - period[0]).total_seconds()
-        return round(ratio, 1)
+        return 100 * 3600 * value / (period[1] - period[0]).total_seconds()
```
People will bring pitchforks if they history get values 0.31232312312312312312312

cgomesu replied:
The rounding operation for `ratio` type sensors was moved from the static method you mentioned to  the `state` property (L208-212).  That is, the ratio value is rounded immediately prior to returning the state value of the sensor.  By default, it will always round to 2 decimals but with the addition of the new `precision:` configuration variable, users can now modify the number of decimals as well.

dgomes commented on homeassistant/components/history_stats/sensor.py:
```diff
@@ -115,7 +149,12 @@ def __init__(
+        # use UNITS dict for all but 'time' sensors, which use the config variable time_unit
+        self._unit_of_measurement = (
+            UNITS[sensor_type] if sensor_type != CONF_TYPE_TIME else time_unit
+        )
```
so UNITS[CONF_TYPE_TIME] is never used, right ?

cgomesu replied:
it is used to validate `CONF_TIME_UNIT` (see L80-82), which is provided by users via the new `time_unit:` configuration variable or defaults to `h` (hours).  please notice that the addition and usage of the `UNITS` dict was not my idea.  I simply adapted the implementation of the new configuration variables to the previous code.

dgomes commented:
@cgomesu PR review take time... it's not something particular about your PR, all PR get the same treatment

cgomesu commented:
no worries, @dgomes .  :+1:   I saw that recent changes introduced conflicts and I just do not have the time to look into this anymore, so I closed the PR.  I also noticed that the introduction of the statistics table in the DB and the long-term stats started moving such analytical integrations towards a different (and better) direction.  Think it's better to build upon that instead.
